components/core/src/clp/ffi/KeyValuePairLogEvent.cpp

```diff
@@ -0,0 +1,174 @@
+        SchemaTree const& schema_tree,
+        KeyValuePairLogEvent::NodeIdValuePairs const& node_id_value_pairs
+) -> std::optional<std::errc>;
+
+/**
+ * @param schema_tree
+ * @param node_id
+ * @param node_id_value_pairs
+ * @return Whether the given node is a leaf node in the sub schema tree defined by
+ * `node_id_value_pairs`. A node is considered a leaf if none of its descendants appear in the
+ * `node_id_value_pairs`.
+ */
+[[nodiscard]] auto is_leaf_node(
+        SchemaTree const& schema_tree,
+        SchemaTreeNode::id_t node_id,
+        KeyValuePairLogEvent::NodeIdValuePairs const& node_id_value_pairs
+) -> bool;
+
+auto node_type_matches_value_type(SchemaTreeNode::Type type, Value const& value) -> bool {
+    switch (type) {
+        case SchemaTreeNode::Type::Obj:
+            return value.is_null();
+        case SchemaTreeNode::Type::Int:
+            return value.is<value_int_t>();
+        case SchemaTreeNode::Type::Float:
+            return value.is<value_float_t>();
+        case SchemaTreeNode::Type::Bool:
+            return value.is<value_bool_t>();
+        case SchemaTreeNode::Type::UnstructuredArray:
+            return value.is<FourByteEncodedTextAst>() || value.is<EightByteEncodedTextAst>();
+        case SchemaTreeNode::Type::Str:
+            return value.is<string>() || value.is<FourByteEncodedTextAst>()
+                   || value.is<EightByteEncodedTextAst>();
+        default:
+            return false;
+    }
+}
+
+auto validate_node_id_value_pairs(
+        SchemaTree const& schema_tree,
+        KeyValuePairLogEvent::NodeIdValuePairs const& node_id_value_pairs
+) -> std::optional<std::errc> {
+    std::optional<std::errc> ret_val;
+    std::unordered_map<SchemaTreeNode::id_t, std::unordered_set<std::string_view>>
+            parent_node_id_to_key_names;
+    try {
+        for (auto const& [node_id, value] : node_id_value_pairs) {
+            if (SchemaTree::cRootId == node_id) {
+                ret_val.emplace(std::errc::operation_not_permitted);
+                break;
+            }
+
+            auto const& node{schema_tree.get_node(node_id)};
+            auto const node_type{node.get_type()};
+            if (false == value.has_value()) {
+                // Value is an empty object (`{}`, which is not the same as `null`)
+                if (SchemaTreeNode::Type::Obj != node_type) {
+                    ret_val.emplace(std::errc::protocol_error);
+                    break;
+                }
+            } else if (false == node_type_matches_value_type(node_type, value.value())) {
+                ret_val.emplace(std::errc::protocol_error);
+                break;
+            }
+
+            if (SchemaTreeNode::Type::Obj == node_type
+                && false == is_leaf_node(schema_tree, node_id, node_id_value_pairs))
+            {
+                // Implicit key conflict: A `null` or empty value is given but its descendants
+                // appear in the node ID value pairs
```

Comment:
```suggestion
                // The node's value is `null` or `{}` but its descendants appear in
                // `node_id_value_pairs`.
```

I think saying implicit/explicit key conflict is a bit more confusing than not.